### PR TITLE
feat(server): SORT command prototype

### LINKED ISSUE #287: Support SORT command
Part of #91

### PR DESCRIPTION
**Pull request by dranikpg**
fixes #287

### CODE REVIEW

File: src/server/container_utils.h [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+// STRING = std::string
+// INTEGRAL = int64_t
+// MIXED = std::variant<std::string, int64_t>
+template <StorageKind K>
+using StorageType =
+    std::conditional_t<K == StorageKind::MIXED, std::variant<std::string, int64_t>,
+                       std::conditional_t<K == StorageKind::STRING, std::string, int64_t>>;
+
+// IsContainer returns true if the iterator points to a container type.
+bool IsContainer(PrimeIterator it) {
```
Comment (romange) [inline src/server/container_utils.h:47]: non template functions require inline in header files

Comment (romange) [reply src/server/container_utils.h:47]: here and everywhere

File: src/server/container_utils.h [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+// Iterate over all values and call func(val). Iteration stops as soon
+// as func return false.
+template <typename F> bool IterateList(PrimeIterator it, F&& func) {
+  quicklist* ql = static_cast<quicklist*>(it->second.RObjPtr());
+  quicklistIter* qiter = quicklistGetIteratorAtIdx(ql, AL_START_HEAD, 0);
+  quicklistEntry entry = QLEntry();
+  bool success = true;
+  while (success && quicklistNext(qiter, &entry)) {
```
Comment (romange) [inline src/server/container_utils.h:85]: it looks that the code could be a bit simpler here.
1. instead of `while`, maybe use `do { } while (success)`.
2. There is no point in boolean AND since it only goes from true to false once. So replace `&=` with `=`.

File: src/server/container_utils.h (hunk 1 above)
Comment (dranikpg) [reply src/server/container_utils.h:47]: This is just a prototype so far to check the general approach. I'll look at details later, once its set

File: src/server/container_utils.h [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  bool success = true;
+  if (encoding == kEncodingIntSet) {
+    intset* is = static_cast<intset*>(ptr);
+    int64_t ival;
+    int ii = 0;
+    char buf[32];
+
+    while (success && intsetGet(is, ii++, &ival)) {
+      if constexpr (std::is_invocable_v<F, int64_t>) {
+        success &= wrapped_func(ival);
```
Comment (romange) [inline src/server/container_utils.h:123]: the same here

File: src/server/container_utils.h [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    }
+  }
+  quicklistReleaseIterator(qiter);
+  return success;
+}
+
+// Iterate over all values and call func(val). If func returns a bool and
+// returns false, iteration stops. func might also return void.
+template <typename F, typename RetV = std::invoke_result_t<F, std::string&&>>
+RetV IterateSet(PrimeIterator it, F&& func) {
```
Comment (dranikpg) [inline src/server/container_utils.h:99]: IterateSet supports void return to replace FillSet from set_family

File: src/server/container_utils.h [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+      if constexpr (std::is_invocable_v<F, int64_t>) {
+        success &= wrapped_func(ival);
+      } else {
+        char* next = absl::numbers_internal::FastIntToBuffer(ival, buf);
+        success &= wrapped_func(std::string{buf, size_t(next - buf)});
+      }
+    }
+  } else {
+    std::string str;
+    if (absl::GetFlag(FLAGS_use_set2)) {
```
Comment (romange) [inline src/server/container_utils.h:131]: you do not need to fetch a flag. use `encoding` for that.

File: src/server/generic_family.cc [hunk 6]
```diff
@@ -463,6 +458,164 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
...
+
+  bool Parse(std::string&& item) {
+    if constexpr (STORE_SCORE) {
+      if (!absl::SimpleAtod(item, &this->score))
+        return false;
+    }
+    if constexpr (std::is_assignable_v<KEY_TYPE&, std::string&&>) {
+      key = std::move(item);
+    } else {
+      assert(false && "unreachable");
```
Comment (romange) [inline src/server/generic_family.cc:497]: can it be `static_assert` ?

File: src/server/generic_family.cc [hunk 7]
```diff
@@ -463,6 +458,164 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
...
+
+template <
+    // Whether alphanumeric sort will be used.
+    bool ALPHA,
+    // Deduced StorageKind type for optimal size.
+    container_utils::StorageKind STORAGE_KIND,
+    // Whether to store double score. It will be stored for non-alphanumeric sorts
+    // when our items are not trivailly convertiable to doubles (i.e. they are not integers)
+    bool STORE_SCORE = !ALPHA && STORAGE_KIND != container_utils::StorageKind::INTEGRAL>
+struct SortEntry : public std::conditional_t<STORE_SCORE, SortEntryScore, SortEntryEmpty> {
```
Comment (dranikpg) [inline src/server/generic_family.cc:475]: I initially tried this with more inheritance. But the logic quickly becomes scattered and it seems to me that `if constexpr()` chains are easier to understand.

File: src/server/generic_family.cc [hunk 8]
```diff
@@ -463,6 +458,164 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
...
+        mixed);
+  }
+};
+
+template <bool ALPHA, container_utils::StorageKind STORAGE_KIND>
+OpStatus SortInternal(PrimeIterator it) {
+  using SENTRY = SortEntry<ALPHA, STORAGE_KIND>;
+  std::vector<SENTRY> sort_entries;
+
+  VLOG(0) << "Storage kind " << int(STORAGE_KIND) << " size " << sizeof(SENTRY);
```
Comment (romange) [inline src/server/generic_family.cc:541]: `DVLOG(1)` probably

File: src/server/container_utils.h [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+      }
+      dictReleaseIterator(di);
+    }
+  }
+  return success;
+}
+
+// Iterate over all values and call func(val). Iteration stops as soon
+// as func returns false.
+template <typename F> bool Iterate(PrimeIterator it, F&& func) {
```
Comment (romange) [inline src/server/container_utils.h:155]: you do not really use `Iterator`. Why not accept `const PrimeValue& obj` instead and call it with `it->second` ?

File: src/server/generic_family.cc (hunk 6 above)
Comment (dranikpg) [reply src/server/generic_family.cc:497]: No, because all code paths are compiled. The `STORAGE_KIND` is just a kind of "agreement" that does not exist on the type level.
It could, however, be part of the type system if we wire it down to the Iterate functions.  

I don't like this part either, but I'm not sure the alternative is worth the effort

File: src/server/generic_family.cc [hunk 10]
```diff
@@ -463,6 +458,164 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
...
+    }
+  }
+};
+
+OpStatus OpSort(const OpArgs& op_args, std::string_view key, bool alpha) {
+  using namespace container_utils;
+
+  auto [it, _] = op_args.shard->db_slice().FindExt(op_args.db_ind, key);
+  if (!IsValid(it) || !IsContainer(it)) {
+    return OpStatus::KEY_NOTFOUND;
```
Comment (romange) [inline src/server/generic_family.cc:600]: for non container redis returns "not found" ?

File: src/server/generic_family.cc (hunk 8 above)
Comment (dranikpg) [reply src/server/generic_family.cc:541]: I'll add normal logs once this is no longer a prototype. This is just for me so far

**Review by romange — COMMENTED** [review]
looks good

File: src/server/generic_family.cc [hunk 11]
```diff
@@ -463,6 +458,209 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
...
+  return SortEntry<ALPHA, container_utils::StorageKind{STORAGE}>{};
+}
+
+template <size_t... seq> static auto ListTypes(std::index_sequence<seq...>) {
+  return std::variant<std::vector<decltype(MapType<seq>())>...>{};
+}
+}  // namespace helper
+
+// std::variant of all possible vector<SortEntry<*, *>>
+using SortEntryList = decltype(helper::ListTypes(std::make_index_sequence<6>{}));
```
Comment (dranikpg) [inline src/server/generic_family.cc:545]: Tricky but works... 😕 

The real issue is that if someone decides to add some more enum variants, then this will break easily and silently. Maybe add a few static_asserts on the constants. AFAIK there is no way to get the enum size (without adding a `__LAST` member) 🤔

File: src/server/generic_family.cc [hunk 12]
```diff
@@ -463,6 +458,209 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
...
+  if (entries.status() == OpStatus::WRONG_TYPE)
+    return (*cntx)->SendError("One or more scores can't be converted into double");
+
+  if (!entries.ok())
+    return (*cntx)->SendNullArray();
+
+  std::visit(
+      [cntx, bounds, reversed](auto& entries) {
+        if (bounds) {
+          std::partial_sort(entries.begin(), entries.begin() + bounds->first + bounds->second,
```
Comment (dranikpg) [inline src/server/generic_family.cc:628]: I don't like the alignment, but this is default vscode formatting

File: src/server/generic_family.cc [hunk 13]
```diff
@@ -463,6 +458,209 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
...
+        entries.reserve(pv.Size());
+        return Iterate(pv, [&entries](auto&& val) {
+          return entries.emplace_back().Parse(std::forward<decltype(val)>(val));
+        });
+      },
+      result);
+  return success ? OpResult{std::move(result)} : OpStatus::WRONG_TYPE;
+}
+
+void GenericFamily::Sort(CmdArgList args, ConnectionContext* cntx) {
```
Comment (dranikpg) [inline src/server/generic_family.cc:587]: This function is almost 80 lines long, maybe it'll be more readable when splitted

**Comment by dranikpg** [comment]
Time to clean up & fix details & add tests

The code structure and approach should be set for now. I replaced all custom dispatch logic with just an `std::variant` because it allows easy type erasing when returning the result from the first hop

**Comment by dranikpg** [comment]
```
Dragonfly: 
fill took 15.705727577209473
list + numeric      : 0.6631001830101013
list + alpha        : 0.1616811752319336
set + numeric       : 1.011125087738037
set + alpha         : 0.7711703181266785

Redis:
fill took 20.41540026664734
list + numeric      : 0.6808633804321289
list + alpha        : 5.502249240875244
set + numeric       : 4.273905873298645
set + alpha         : 16.4713973402977
```

This are my benchmarks for far. 

* The list + numeric benchmarks are on par as expected
* set + numeric is about 4 times faster which also is expected because of the additional `StorageType::INTEGRAL` optimizations
* I have no plausible explanation for the alphanumeric tests... I double checked for correct results. This should be the slowest one because we allocate new strings whereas redis does not. Maybe `std::string` compare is faster because it can tell the string size beforehand? Do you have any assumptions?

File: src/server/generic_family.cc [hunk 14]
```diff
@@ -463,6 +458,209 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
...
+    if constexpr (!ALPHA && STORAGE_KIND == container_utils::StorageKind::MIXED)
+      std::visit(std::forward<F>(f), key);
+    else
+      f(key);
+  }
+};
+
+namespace helper {
+// Sort entry indices are encoded with the formula ALPHA * 3 + STORAGE
+template <size_t I, bool ALPHA = I / 3, size_t STORAGE = I % 3> static auto MapType() {
```
Comment (romange) [inline src/server/generic_family.cc:535]: I lost you from here and onward

File: src/server/generic_family.cc (hunk 11 above)
Comment (romange) [reply src/server/generic_family.cc:545]: I would like to understand in more detail what you do here. 
what's this vector does.

File: src/server/generic_family.cc [hunk 15]
```diff
@@ -463,6 +458,209 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
...
+    // Whether to store double score. It will be stored for non-alphanumeric sorts
+    // when our items are not trivailly convertiable to doubles (i.e. they are not integers)
+    bool STORE_SCORE = !ALPHA && STORAGE_KIND != container_utils::StorageKind::INTEGRAL>
+struct SortEntry : public std::conditional_t<STORE_SCORE, SortEntryScore, SortEntryEmpty> {
+  // Key type. If we sort alphanumerically, we store strings. Else we mimic
+  // the container storage type.
+  using KEY_TYPE =
+      std::conditional_t<ALPHA, std::string, container_utils::StorageType<STORAGE_KIND>>;
+  // Comparison type. std:string for alphanumeric sort, double otherwise.
+  using COMPARE_TYPE =
```
Comment (romange) [inline src/server/generic_family.cc:481]: when your runtime parameter alpha is true, then STORE_SCORE is false, your SortEntry has only a string.
this branch you can do at the beginning and I think if you do that, you kinda consolidate all the conditionals you put in all the levels. you still write the dedicated code to sort strings but you write it here as well, but now it's just divided between all these templates.

but if you just do this one `if` at the upper level for `alpha=true`, you probably end up with less lines of code and at least alpha=true branch will be much clearer. What do you think?

File: src/server/container_utils.h [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+#include "core/compact_object.h"
+#include "core/string_set.h"
+#include "server/common.h"
+
+namespace dfly {
+
+namespace container_utils {
+
+// StorageKind specifies how the container stores values.
+enum class StorageKind {
```
[the remark this answers is not in the document]
Comment (dranikpg) [reply src/server/container_utils.h:27]: No. Should I move them back to generic_family?

File: src/server/container_utils.h [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+    }
+  }
+  quicklistReleaseIterator(qiter);
+  return success;
+}
+
+// Iterate over all values and call func(val). If func returns a bool and
+// returns false, iteration stops. func might also return void.
+template <typename F, typename RetV = std::invoke_result_t<F, std::string&&>>
+RetV IterateSet(const PrimeValue& pv, F&& func) {
```
[the remark this answers is not in the document]
Comment (dranikpg) [reply src/server/container_utils.h:116]: This would mean we promote everything to a string. This would also mean dropping support for Integral container types. However given that integer sets are limited to small sizes this would probably not make a significant impact altogether.

File: src/server/container_utils.h [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+      dictEntry* de = nullptr;
+      while (success && (de = dictNext(di))) {
+        str.assign((sds)de->key, sdslen((sds)de->key));
+        success = wrapped_func(std::move(str));
+      }
+      dictReleaseIterator(di);
+    }
+  }
+
+  if constexpr (std::is_same_v<RetV, bool>)
```
[the remark this answers is not in the document]
Comment (dranikpg) [reply src/server/container_utils.h:165]: Why not? It works alright

File: src/core/compact_object.cc [hunk 19]
```diff
@@ -262,6 +262,8 @@ size_t RobjWrapper::Size() const {
     case OBJ_STRING:
       DCHECK_EQ(OBJ_ENCODING_RAW, encoding_);
       return sz_;
+    case OBJ_LIST:
```
[the remark this answers is not in the document]
Comment (dranikpg) [reply src/core/compact_object.cc:265]: To call `reserve(Primvevalue::Size())` on the entries vector

File: src/server/generic_family.cc (hunk 11 above)
Comment (dranikpg) [reply src/server/generic_family.cc:545]: I collected all possible vector<SortEntry<ALPHA, STORAGE>> instantiations into a `std::variant<>`

File: src/server/container_utils.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (romange) [reply src/server/container_utils.h:27]: i think it will be better

File: src/server/container_utils.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (romange) [reply src/server/container_utils.h:116]: ah, I missed this.

[the remark this answers is not in the document]
Comment (romange) [reply src/server/container_utils.h:116]: though you use generics but then there is a run-time dispatch inside.

File: src/server/generic_family.cc (hunk 11 above)
Comment (romange) [reply src/server/generic_family.cc:545]: you try to write a generic code that manages "generic" data structures and by generic I mean - union of vectors.
You say your type space is ALPHAxSTORAGE, but 
1. can ALPHA=TRUE and STORAGE be non-string? 
2. when ALPHA is FALSE you have list data structure or set with non-int encoding, your storage is STRING? 

what I am trying to say is that your contention point is probably the sorting algorithm where you want to have the most optimized code working on a compile-known type. Everything else can be deduced at run-time. So why not decompose your states in such way that they will lead into 3 branches: sort by string, sort by int and sort by MIXED. But once you are inside the branch you just write the code for that branch. 
Maybe I am oversimplifying things - tell me if I am wrong..

File: src/server/generic_family.cc (hunk 15 above)
Comment (dranikpg) [reply src/server/generic_family.cc:481]: It will in this case likely.

However where to divide? A full dedicated branch? So... sort & write is duplicated? Might be okay for now.

Later (in a second PR) SORT will be extended with external keys and there it'll be extremely inconvenient to have multiple branches because lots of hard logic will be duplicated. So there should be at least some `variant<vector<T?>>` that can be interfaced with from common code so that:
* T? can return its key
* T? will possibly store an external compare by value
* T? can be compared (without overhead for numeric sort)

We can drop support for integers at all because integer encoding is only used in small containers. It'll be beneficial only if you sort a container thousands of times... which is an unlikely use-case.

It could be even simpler, we can drop support for partial integer encoding (the `variant<string,int>` type). However, we should keep `Iterate` templated, because in case of a numerical sort we could convert a long into a string and then parse that string as a double 🤯 Parsing doubles is really slow... Converting integers to strings is probably not, because it fits into SSO.

This leaves us with:
- [string] for alphanumeric sort without external keys
- [string, double] for numeric sort with or without external keys
- [string, string] for alphanumeric sort with external key ** out of scope for this PR **

We could introduce just one variable - ALPHA - so we keep the conditional inheritance and Cmp functions (the `: conditional<ALPHA, Score, Empty>` part). Or even one common type [string, string, double] (key, cmp_key, cmp_double) that covers all cases for the long and short run ([string, double] in this PR). However we could no longer be on par with redis sort performance

Comment (romange) [reply src/server/generic_family.cc:481]: 1. does `vector` represent a single item to sort in case of multiple GET statements (aka `sort s1 by w_* get w_* get f_* GET #`)? Or vector represents the list of items you sort?  

2. What's this "partial integer encoding" ? I am having a hard time correlating this name with SORT documentation.
3. I am inclined to drop the support for integers if it makes the code simpler.
4. Regarding SortEntry design. I do not understand why `key` is either string, int or a variant. which option in the spec the variant handles? What does it mean having a key and a score? to which option it correlates in the spec? 
5. 
`T? can return its key` - is it the original key (int, or string) covering SET, LIST, or ZSET field names?
`T? will possibly store an external compare by value` is it for `BY` statement, so BY fetches string or double (Controlled by ALPHA)? 
`T? can be compared (without overhead for numeric sort)`: you are referring to non-alpha case where your compare operator has a scalar (double) or a pair<by_score, string> or for alpha case where you compare a scalar (string) or a `pair<by_string, key_string>` ?

File: src/server/generic_family.cc (hunk 11 above)
Comment (romange) [reply src/server/generic_family.cc:545]: In any case, the sequence generation is not justified here.
It's artificial in order to save a few lines. Lets use the explicit variant with comments mapping each type to a combination of SORT sub-case.  I am still struggling to see whether you really need the whole cartesian product of possibilities.

File: src/server/generic_family.cc [hunk 20]
```diff
@@ -463,6 +462,170 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
...
+      if (!absl::SimpleAtoi(ArgS(args, i + 1), &offset) ||
+          !absl::SimpleAtoi(ArgS(args, i + 2), &limit)) {
+        return (*cntx)->SendError(kInvalidIntErr);
+      }
+      bounds = {offset, limit};
+      i += 2;
+    }
+  }
+
+  auto entries = cntx->transaction->ScheduleSingleHopT([&](Transaction* t, EngineShard* shard) {
```
Comment (romange) [inline src/server/generic_family.cc:590]: I suspect that entries are of type `SortEntryList`. Can you pls make it explicit?

File: src/server/generic_family.cc [hunk 21]
```diff
@@ -463,6 +462,170 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
...
+  });
+
+  if (entries.status() == OpStatus::WRONG_TYPE)
+    return (*cntx)->SendError("One or more scores can't be converted into double");
+
+  if (!entries.ok())
+    return (*cntx)->SendNullArray();
+
+  std::visit(
+      [cntx, bounds, reversed, key](auto& entries) {
```
Comment (romange) [inline src/server/generic_family.cc:601]: pls define the lambda before calling the function. it's a bit hard to follow where its body ends.

**Review by romange — COMMENTED** [review]
I think it looks MUCH better.
is it still draft?

**Comment by dranikpg** [comment]
I'll make it non draft as soon as I add tests

File: src/server/generic_family_test.cc [hunk 22]
```diff
@@ -169,7 +169,7 @@ TEST_F(GenericFamilyTest, Stick) {
   // check stick returns zero on non-existent keys
   ASSERT_THAT(Run({"stick", "a", "b"}), IntArg(0));
 
-  for (auto key: {"a", "b", "c", "d"}) {
+  for (auto key : {"a", "b", "c", "d"}) {
     Run({"set", key, "."});
```
Comment (dranikpg) [inline src/server/generic_family_test.cc:173]: vscode formatter

File: src/server/generic_family.cc [hunk 23]
```diff
@@ -463,6 +463,170 @@ void GenericFamily::Stick(CmdArgList args, ConnectionContext* cntx) {
...
+  OpResult<SortEntryList> entries =
+      cntx->transaction->ScheduleSingleHopT([&](Transaction* t, EngineShard* shard) {
+        return OpFetchSortEntries(t->GetOpArgs(shard), key, alpha);
+      });
+
+  if (entries.status() == OpStatus::WRONG_TYPE)
+    return (*cntx)->SendError("One or more scores can't be converted into double");
+
+  if (!entries.ok())
+    return (*cntx)->StartArray(0);
```
Comment (dranikpg) [inline src/server/generic_family.cc:600]: Maybe add `SendEmptyArray()` to redisbuilder? (null != empty)

Comment (romange) [reply src/server/generic_family.cc:600]: sure, also write there in the comments the distinction in the output.

File: src/server/generic_family_test.cc [hunk 24]
```diff
@@ -276,4 +276,59 @@ TEST_F(GenericFamilyTest, Scan) {
...
+  ASSERT_THAT(Run({"sort", "list-1", "LIMIT", "5", "2"}), ArrLen(0));
+  // limits desc
+  ASSERT_THAT(Run({"sort", "list-1", "DESC", "LIMIT", "0", "5"}).GetVec(), ElementsAre("200", "10.1", "3.5", "2.20", "1.2"));
+  ASSERT_THAT(Run({"sort", "list-1", "DESC", "LIMIT", "2", "2"}).GetVec(), ElementsAre("3.5", "2.20"));
+  ASSERT_THAT(Run({"sort", "list-1", "DESC", "LIMIT", "1", "1"}), "10.1");
+  ASSERT_THAT(Run({"sort", "list-1", "DESC", "LIMIT", "5", "2"}), ArrLen(0));
+
+  // Test set sort
+  Run({"del", "set-1"});
+  Run({"lpush", "set-1", "5.3", "4.4", "60", "99.9", "100", "9"});
```
Comment (romange) [inline src/server/generic_family_test.cc:306]: it's not set :)

**Review by romange — COMMENTED** [review]
lgtm

File: src/server/generic_family.cc (hunk 23 above)
Comment (dranikpg) [reply src/server/generic_family.cc:600]: In reply builder?

I'm not sure what the difference is. I thought of writing that null represents failures/timeouts, but sort shows its not always the case.

**Comment by romange** [comment]
When using each one of them is not upto us at this point. I meant the
protocol difference - null returns *-1 and empty returns *0 items

File: src/core/compact_object.cc [hunk 25]
```diff
@@ -258,10 +258,20 @@ size_t RobjWrapper::MallocUsed() const {
 }
 
 size_t RobjWrapper::Size() const {
+  robj self{.type = type_,
```
Comment (romange) [inline src/core/compact_object.cc:261]: nit: you use it only for zset. I would move it inside the block there

**Review by romange — APPROVED** [review]

**Comment by romange** [comment]
something has changed - it has conflicts with the main branch

**Review by romange — APPROVED** [review]

**Review by romange — APPROVED** [review]

**Review by romange — APPROVED** [review]
